ext/standard/array.c

```diff
@@ -349,6 +350,102 @@ static zend_always_inline int php_array_data_compare_unstable_i(Bucket *f, Bucke
 }
 /* }}} */
 
+/* return int to be compatible with compare_func_t */
+static int hash_zval_strict_function(zval *z1, zval *z2) /* {{{ */
+{
+	ZVAL_DEREF(z1);
+	ZVAL_DEREF(z2);
+	// The most important thing about this comparison mode is that the result
+	// is 0 when zend_is_identical, and non-zero otherwise.
+	if (zend_is_identical(z1, z2)) {
```

Comment:
As you're enumerating all cases below anyway, I'd drop this check and handle equality below as well. For arrays that would save one deep comparison.

Reply:
Yes, but it makes it harder to maintain the invariant, and less obvious if there's some subtle bug somewhere that causes the output of zend_is_identical and this function to diverge.  It's easier (IMO) to determine by inspection that the only way a zero can be returned from this function is if zend_is_identical returns zero.  If you're worried about performance I could probably exclude objects and arrays from this comparison -- arrays involve a call to zend_hash_compare anyway so the "obvious by inspection it can't return 0" thing isn't as obviously true there anyway.

Reply:
(avoided the deep comparison in latest patch, as requested.)